### PR TITLE
EIP 3529: alternative refund reduction

### PR DESCRIPTION
**Pull request by vbuterin**
An alternative way to reduce refunds to deal with the issues addressed in EIPs 3298 and 3403.

### CODE REVIEW

**Comment by alita-moore** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):


	 - File with name EIPS/eip-3529.md is new and new files must be reviewed

**Review by holiman — APPROVED** [review]
LGTM

File: EIPS/eip-3529.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+For blocks where `block.number >= FORK_BLOCK`, the following changes apply.
+
+1. Remove the `SELFDESTRUCT` refund.
+2. Replace `SSTORE_CLEARS_SCHEDULE` with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of EIP 2929 + 2930)
+3. Reduce the max gas refunded after a transaction to `gas_used // NEW_MAX_REFUND_QUOTIENT`
+
+## Rationale
+
+In [EIP 2200](https://eips.ethereum.org/EIPS/eip-2200#specification), three cases for refunds were introduced:
```
Comment (MicahZoltu) [inline EIPS/eip-3529.md:41]:
```suggestion
In [EIP-2200](./eip-2200.md#specification), three cases for refunds were introduced:
```

File: EIPS/eip-3529.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+However, fully preserving refunds in the `new = original = 0 != current` case, and keeping _some_ refund in the other `nonzero -> zero` cases, ensures that a few key use cases that receive (and deserve) favorable gas cost treatment continue to do so. For example, `zero -> nonzero -> zero` storage set patterns continue to cost only ~100 gas. Two important examples of such patterns include:
+
+* Anti-reentrancy locks (typically flipped from 0 to 1 right before a child call begins, and then flipped back to 0 when the child call ends) 
+* ERC20 approve-and-send (the "approved value" goes from zero to nonzero when the token transfer is approved, and then back to zero when the token transfer processes)
+
+### Effect on storage clearing incentives
+
+A criticism of earlier refund removal EIPs ([EIP 3298](https://eips.ethereum.org/EIPS/eip-3298) and [EIP 3403](https://eips.ethereum.org/EIPS/eip-3403)) is that these EIPs fully remove the incentive to set a value to zero, encouraging users to not fully clear a storage slot if they expect even the smallest probability that they will want to use that storage slot again.
```
Comment (MicahZoltu) [inline EIPS/eip-3529.md:71]:
```suggestion
A criticism of earlier refund removal EIPs ([EIP-3298](./eip-3298.md) and [EIP-3403](./eip-3403.md)) is that these EIPs fully remove the incentive to set a value to zero, encouraging users to not fully clear a storage slot if they expect even the smallest probability that they will want to use that storage slot again.
```

File: EIPS/eip-3529.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+If this EIP were to be implemented, a refund of 4800 gas would remain, so there would only be an incentive to keep a storage slot nonzero if you expect a probability of more than `4800 / 17100 = 28.1%` that you will use that slot again. This is not perfect, but it is likely higher than the average person's expectations of later re-acquiring a token with the same address if they clear their entire balance of it.
```
Comment (MicahZoltu) [inline EIPS/eip-3529.md:75]: EIPs should be written for a target audience of future readers after it has been decided that it is desirable.
```suggestion
A refund of 4800 gas remains, so there is only be an incentive to keep a storage slot nonzero if you expect a probability of more than `4800 / 17100 = 28.1%` that you will use that slot again. This is not perfect, but it is likely higher than the average person's expectations of later re-acquiring a token with the same address if they clear their entire balance of it.
```

File: EIPS/eip-3529.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  - Istanbul: 1612
+  - Berlin: 212
+  - EIP-3403: 1112
+  - This EIP: 212
+- '1-2-1':
+  - Istanbul: 1612
+  - Berlin: 212
+  - EIP-3403: 3012
+  - This EIP: 212
+
```
Comment (MicahZoltu) [inline EIPS/eip-3529.md:143]: This isn't a valid EIP section.  Consider converting this into a test case, or putting it as a subheading in the Rationale section.
```suggestion
```

File: EIPS/eip-3529.md (hunk 2 above)
Comment (MicahZoltu) [reply EIPS/eip-3529.md:71]: EIP-3298 and 3403 are unlikely to be finalized before this one, which means referencing them will block this EIP from moving to final.  Recommend removing references to them and instead just vaguely discuss alternative proposals directly here like "Another option would be to do X, but that would have problem Y."

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Only blockers are the external EIP links and the non-standard section.  The other points of feedback are not required for Draft.

**Review by alita-moore — COMMENTED** [review]
testing bot

File: EIPS/eip-3529.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-3298-removal-of-refunds/5430
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-22
+---
+
+## Simple Summary
+
+Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE.
```
Comment (vbuterin) [inline EIPS/eip-3529.md:14]:
```suggestion
Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE to a lower level where the refunds are still substantial, but they are no longer high enough for current "exploits" of the refund mechanism to be viable.
```

File: EIPS/eip-3529.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
+---
+eip: 3529
+title: Reduction in refunds
+author: Vitalik Buterin (@vbuterin), Martin Swende (@holiman)
+discussions-to: https://ethereum-magicians.org/t/eip-3298-removal-of-refunds/5430
```
Comment (vbuterin) [inline EIPS/eip-3529.md:5]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-3529-reduction-in-refunds-alternative-to-eip-3298-and-3403-that-better-preserves-existing-clearing-incentives/6097
```

**Review by alita-moore — COMMENTED** [review]
test bot

File: EIPS/eip-3529.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+| Constant | Value |
+| - | - |
+| `FORK_BLOCK` | TBD |
+| `NEW_MAX_REFUND_QUOTIENT` | 5 |
+
+For blocks where `block.number >= FORK_BLOCK`, the following changes apply.
+
+1. Remove the `SELFDESTRUCT` refund.
+2. Replace `SSTORE_CLEARS_SCHEDULE` with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```
Comment (vbuterin) [inline EIPS/eip-3529.md:36]:
```suggestion
2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2929.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```

File: EIPS/eip-3529.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+| Constant | Value |
+| - | - |
+| `FORK_BLOCK` | TBD |
+| `NEW_MAX_REFUND_QUOTIENT` | 5 |
+
+For blocks where `block.number >= FORK_BLOCK`, the following changes apply.
+
+1. Remove the `SELFDESTRUCT` refund.
+2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2929.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```
Comment (vbuterin) [inline EIPS/eip-3529.md:36]:
```suggestion
2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2200.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```

File: EIPS/eip-3529.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-3529-reduction-in-refunds-alternative-to-eip-3298-and-3403-that-better-preserves-existing-clearing-incentives/6097
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-22
+---
+
+## Simple Summary
+
+Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE to a lower level where the refunds are still substantial, but they are no longer high enough for current "exploits" of the refund mechanism to be viable.
```
Comment (MicahZoltu) [inline EIPS/eip-3529.md:14]: Recommend shortening this up.  Think email subject line or forum thread title or GitHub PR title.
```suggestion
Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE.
```
The latter part of this sentence should be left for the motivation section.

**Review by MicahZoltu — APPROVED** [review]
One more suggestion, but not a blocker.

**Comment by abo-L** [comment]
> An alternative way to reduce refunds to deal with the issues addressed in EIPs 3298 and 3403.

File: EIPS/eip-3529.md (hunk 6 above)
Comment (abo-L) [reply EIPS/eip-3529.md]: @ABOLFAZL4343

**Pull request merged**